### DISCUSSION TITLE
Storing user profile's data on Blockchain

### DISCUSSION THREAD

**Discussion by qualwebs** [bitshares/bitshares-core#2811] [Q&A]
How can I allow the user to store its profile's data on blockchain ? **My use case goes something like this :-**

I want to store user profile's data like username, age, email and related fields on blockchain itself. However, the problem is when I try to make an API Call to store data, I get a `missing_active_authority` error.

Example : Based on my observations, If I am trying to store data inside nathan's account, then nathan's account must satisfy following two conditions inside cli_wallet :-

   1. Nathan's Private keys must be imported inside the `Cli_Wallet` using the `import_key account_name [private_key]` command.
   2. Nathan's Account must have sufficient amount of funds to perform custom operation (Store data) 

Example : 
1. User will have the following form to store the data, and the username will be the account's name.

![Screenshot from 2024-03-05 16-41-01](https://github.com/bitshares/bitshares-core/assets/13719000/7cfccadf-46ff-44b3-bf66-070c34d8b0ec)

2. When I click submit, a node API is hit that is connected with the CLI Wallet's API and calls the `account_store_map` function. But, the CLI Wallet is giving the missing required active authority error.  

Missing Required Active Authority Error :-
![Screenshot from 2024-03-05 16-40-23](https://github.com/bitshares/bitshares-core/assets/13719000/2737c2b5-b0bf-41cb-ae30-70172afb74f2)

I wanted to ask, is there any other way to allow user to store data? How a typical user on web wallet perform the operations, why don't they get this error and is there anyway to achieve something similar to this using an API built from scratch ?
The above operation can be easily done through `CLI Wallet` but if I am implementing this functionality in my app, then a typical user will not be using the cli wallet right ? That's why I want to explore if this would be possible, any guidance and help will be greatly appreciated from the community.

**Comment by abitmore** [comment]
>1. Nathan's Private keys must be imported inside the Cli_Wallet using the import_key account_name [private_key] command.
>2. Nathan's Account must have sufficient amount of funds to perform custom operation (Store data)

3. The wallet must be unlocked.

**Comment by abitmore** [comment]
> When I click submit, a node API is hit that is connected with the CLI Wallet's API

Please be aware that web wallet connects to API nodes directly, just like cli_wallet (which connects to an API node too). And web wallet manages private keys too. In this case, you need to sign the transaction (which contains the "custom_operation") in the web wallet (just like what cli_wallet does), then broadcast the signed transaction to the network via the API node, but not call a **wallet API** in the web wallet - you should not call "wallet API" in web wallet but only "node APIs". See https://dev.bitshares.works/en/master/development/apps/network_general.html for network structure.

**Comment by qualwebs** [comment]
> In this case, you need to sign the transaction (which contains the "custom_operation") in the web wallet

By signing the transaction in the web wallet, you mean I need to modify the web wallet code to sign a transaction in it by using some library like bitsharesjs ?

Also, for signing a transaction I will need a private key but when the user logs-in or creates an account from the web wallet they only have a random password generated by browser. Do you mean to say something like?.... - "I should extract the private key from other part of the codebase of the web wallet (bitshares-ui) in my Update Profile Form, and then sign a transaction in this form itself".


> you should not call "wallet API" in web wallet but only "node APIs

Another thing, the node api as I saw doesn't have the `account_store_map` method, then how will I be able to store the data ?

**Reply by abitmore** [reply]
Since it looks like you're already adding the "Update Profile Form" in `bitshares-ui`, I think serialization of `account_storage_map` need to be implemented in `bitsharesjs` so that it can be used in `bitshares-ui`. Note that `account_storage_map` might be a little different than other operations there because it is wrapped in `custom_operation`. BTW you don't need to worry about how to get the private key from the password in `bitshares-ui`, usually you only need to call the `sign()` function or alike if all the required work in `bitsharesjs` is done.

**Reply by abitmore** [reply]
> how will I be able to store the data

You store data on the blockchain by signing a transaction and broadcasting it to the network (then it gets included into a new block).

**Comment by qualwebs** [comment]
> Since it looks like you're already adding the "Update Profile Form" in `bitshares-ui`

Indeed, yes.

> I think serialization of `account_storage_map` need to be implemented in `bitsharesjs`

I explored the code base of bitsharesjs, found some piece of code that's implementing the various different operations, I've defined a temporary operation for storing data `account_store_map` inside operations.js, (will have to modify this according to `account_storage_map`'s structure right ?) and inorder to implement the serialization for this operation based on your response I guess I have to implement the serialization part in some other separate file, probably under the serialization folder. But what I couldn't wrap my head around is, how do I get the `account_storage_map` accessible in bitsharesjs because it is a part of `custom_operations` plugin which cannot be installed using npm. I wonder whether it will work if I simply copy the plugins inside the bitsharesjs folder and then import it inside my serialization file.

(See the last operation below comments)
![Screenshot from 2024-03-07 15-25-30](https://github.com/bitshares/bitshares-core/assets/13719000/7abd7e67-cf1f-4813-bfc1-6d86df64634f)


Structure of `account_storage_map` :- 
```
struct account_storage_map : chain::base_operation
{
   bool remove;
   string catalog;
   flat_map<string, optional<string>> key_values;

   void validate()const;
};
```


> BTW you don't need to worry about how to get the private key from the password in `bitshares-ui`

Got it, thanks.

**Reply by abitmore** [reply]
Disclaimer: I'm not a JS expert, so my answer here could be wrong.

By adding / defining a new `Serializer` for `account_store_map` in `bitsharesjs`, it becomes a part of the `bitsharesjs` library, then it can be used just like other types or operations. Nevertheless, it is not directly a part of an operation, but a structure to describe data in the `data` field of `custom_operation`, so it is a challenge - how to add it so that it works and how to use it. Not like other types, I guess we may need to explicitly call `serialize()` somewhere.

Please check examples in https://github.com/bitshares/bitsharesjs/tree/develop/examples for how to use and test the new structure.

BTW I think you need to modify the source code of `bitsharesjs` and then rebuild it, then install it (E.G. with `npm`) locally for development, but not directly editing code in `node_modules/*/dist`, because the code might get overwritten accidentally under some circumstances. For production use you need to submit your code to Github then create a pull request at https://github.com/bitshares/bitsharesjs/pulls .

**Reply by qualwebs** [reply]
So, to summarize what you mean is I have to do following things so that I might be able to store data in the user's account :-

1. Somehow figure out how to import the `custom_operations_plugin` from the `bitshares-core/lib/plugins/custom_operations` into the `bitsharesjs` and making it work is a challenge
2. Define my own custom object structure for `account_storage_map` (that will be available now because we imported `custom_operations_plugin` in previous step)
3. Define a custom serialization method to serialize a new operation that can be used to store data
4. And after all the modification is done, I gotta rebuild it.

**Reply by abitmore** [reply]
Maybe you misunderstood something or I misunderstood you, but we don't import a C++ plugin (`custom_operations_plugin`) into a javascript library (`bitsharesjs`), so your summary looks strange. There was a tool that we used to generate javascript code (for `bitsharesjs`) from C++ code (`bitshares-core`), but that tool is a bit outdated now and I guess it no longer works.
